canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+    {
+        cout << "Canary has reached end of run time" << endl;
+        g_main_loop_quit(data->main_loop);
+    }
+
+    // If intermittent run, check if Canary should be paused
+    if(STRCMP(data->pCanaryConfig->canaryRunScenario.c_str(), "Intermittent") == 0 && 
+        duration_cast<minutes>(system_clock::now().time_since_epoch()).count() > data->runTill)
+    {
+        data->timeOfNextKeyFrame->clear();
+        int sleepTime = ((rand() % 10) + 1); // [minutes]
+        cout << "Intermittent sleep time is set to: " << sleepTime << " minutes" << endl;
+        data->sleepTimeStamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count(); // [milliseconds]
+        sleep(sleepTime * 60); // [seconds]
+        int runTime = (rand() % 10) + 1; // [minutes]
+        cout << "Intermittent run time is set to: " << runTime << " minutes" << endl;
+        // Set runTill to a new random value 1-10 minutes into the future
+        data->runTill = duration_cast<minutes>(system_clock::now().time_since_epoch()).count() + runTime; // [minutes]
+    }
+
+CleanUp:
+
+    if (info.data != nullptr) {
+        gst_buffer_unmap(buffer, &info);
+    }
+
+    if (sample != nullptr) {
+        gst_sample_unref(sample);
+    }
+
+    return ret;
+}
+
+// This function is called when an error message is posted on the bus
+static void error_cb(GstBus *bus, GstMessage *msg, CustomData *data) {
+    GError *err;
+    gchar *debug_info;
+
+    // Print error details on the screen
+    gst_message_parse_error(msg, &err, &debug_info);
+    g_printerr("Error received from element %s: %s\n", GST_OBJECT_NAME (msg->src), err->message);
+    g_printerr("Debugging information: %s\n", debug_info ? debug_info : "none");
+    g_clear_error(&err);
+    g_free(debug_info);
+
+    g_main_loop_quit(data->main_loop);
+}
+
+void kinesis_video_init(CustomData *data) {
+    unique_ptr<DeviceInfoProvider> device_info_provider(new CanaryDeviceInfoProvider());
+    unique_ptr<ClientCallbackProvider> client_callback_provider(new CanaryClientCallbackProvider());
+    unique_ptr<StreamCallbackProvider> stream_callback_provider(new CanaryStreamCallbackProvider
+(reinterpret_cast<UINT64>(data)));
+
+    char const *accessKey;
+    char const *secretKey;
+    char const *sessionToken;
+    char const *defaultRegion;
+    string defaultRegionStr;
+    string sessionTokenStr;
+
+    char const *iot_get_credential_endpoint;
+    char const *cert_path;
+    char const *private_key_path;
+    char const *role_alias;
+    char const *ca_cert_path;
+
+    unique_ptr<CredentialProvider> credential_provider;
+
+    if (nullptr == (defaultRegion = getenv(DEFAULT_REGION_ENV_VAR))) {
+        defaultRegionStr = DEFAULT_AWS_REGION;
+    } else {
+        defaultRegionStr = string(defaultRegion);
+    }
+    LOG_INFO("Using region: " << defaultRegionStr);
+
+    if (nullptr != (accessKey = getenv(ACCESS_KEY_ENV_VAR)) &&
```

Comment:
Think if can push this credential configuration within CanaryConfig.cpp. The purpose CanaryConfig.cpp should be to do this kind of env based config set up. Both static and IoT credential provider. See if it can be done as part of `initConfigWithEnvVars`

Reply:
I moved all of the GETENV calls to CanaryConfig.cpp which cleaned it up some, but the rest of the function must stay outside of the CanaryConfig class because it involves CustomData members, and CanaryConfig does not have ability to access CustomData (only vis versa), I can make them point to each other if that would be better.